dora/src/bytecode/generator.rs

```diff
@@ -388,24 +390,47 @@ impl<'a, 'ast> AstBytecodeGen<'a, 'ast> {
             Register::zero()
         };
         self.gen.set_position(expr.pos);
+
         let arg_start_reg = match *call_type {
-            CallType::CtorNew(ty, _) => {
+            CallType::CtorNew(_, _) | CallType::Method(_, _, _) => 1,
+            _ => 0,
+        };
+
+        for (idx, arg) in expr.args.iter().enumerate() {
+            let arg_reg = start_reg.offset(idx + arg_start_reg);
+            self.visit_expr(arg, DataDest::Reg(arg_reg));
+        }
+
+        match *call_type {
+            CallType::CtorNew(_, _) => {
+                let ty = fct.params_with_self()[0];
+                let ty = specialize_for_call_type(&call_type, ty, self.vm);
+                let ty = self.specialize_type(ty);
                 let cls_id = specialize_class_ty(self.vm, ty);
-                self.gen.emit_new_object(start_reg, cls_id);
-                start_reg.offset(1)
+
+                let cls = self.vm.class_defs.idx(cls_id);
+                let cls = cls.read();
+
+                match cls.size {
+                    InstanceSize::Fixed(_) => {
+                        self.gen.emit_new_object(start_reg, cls_id);
+                    }
+                    InstanceSize::Array(_) => {
+                        let length_arg = start_reg.offset(1);
+                        self.gen.emit_new_array(start_reg, cls_id, length_arg);
+                    }
+                    _ => {
+                        unimplemented!();
+                    }
+                }
             }
             CallType::Method(_, _, _) => {
                 let obj_expr = expr.object().expect("method target required");
                 self.visit_expr(obj_expr, DataDest::Reg(start_reg));
-                start_reg.offset(1)
             }
-            _ => start_reg,
+            _ => {}
         };
 
-        for (idx, arg) in expr.args.iter().enumerate() {
```

Comment:
I think this would change the evaluation order of the arguments, consider the following program:

```
class Foo {
  println("Foo");
  fun bar(x: Int) { println("bar"); }
}
fun main() {
  Foo(1).bar(someInt())
}
fun someInt() {
  println("someInt");
}
```

Note this would be `CallType::Method` and not `CallType::Ctor`, so before this change we would have "Foo.someInt.bar", after it would print "someInt.Foo.bar". For `CallType::Method` we need to evaluate the first argument before the others. I guess we should add a test for this, such that we can ensure ast baseline works correctly and cannon matches this behavior.

On the other hand with `CallType::CtorNew` we can move `NewObject` after the arguments. Simply because that's guaranteed to be a simple object allocation (`NewObject` just allocates a raw object and doesn't run the constructor or any arbitrary code). Ast-baseline does this as well, I think the only reason for this is that the arguments could throw/fail and in that case we would have allocated the object for no reason. I think we could also allocate the object and array before the other arguments if this helps you.

On the other hand `visit_expr_call` already got quite large, it might make sense to split it:

1. arrays (we could make use of the fact there will only ever be one argument: length IIRC)
2. CtorNew (construct new object)
3. regular method calls (would probably also contain the parent ctor call)

I think it's better to have a bit of code duplication then to try to force everything through the same method and making the logic more and more complex. Especially with Cannon since generating bytecode is much easier.

I've realized that we already have baseline and cannon already differ in case the object is nil. Baseline will throw immediately and not evaluate the arguments, while cannon will first evaluate all arguments and then check if the object argument is nil. I think cannon is right here though and baseline should change.

Reply:
> Note this would be `CallType::Method` and not `CallType::Ctor`, so before this change we would have "Foo.someInt.bar", after it would print "someInt.Foo.bar". For `CallType::Method` we need to evaluate the first argument before the others.

You're right, thanks for the example. I will change the evaluation order, so that the first argument is always evaluated before the other arguments, regardless of the call type.

>  I think we could also allocate the object and array before the other arguments if this helps you.

I think it is easier, if all arguments are evaluated in one go, but in the end, it doesn't really matter to me. Important for this PR is only, that the length argument is evaluated before `NewArray` is generated.

> On the other hand visit_expr_call already got quite large, it might make sense to split it:
> 1. arrays (we could make use of the fact there will only ever be one argument: length IIRC)
> 2. CtorNew (construct new object)
> 3. regular method calls (would probably also contain the parent ctor call)

I agree with you. After I'm done with this PR, I will create a new PR for this.

Reply:
Done. Added your example as a cannon test to assure the correct evaluation order.